src/main/java/org/springframework/data/elasticsearch/core/ElasticsearchEntityMapper.java

```diff
@@ -0,0 +1,676 @@
+import org.springframework.data.elasticsearch.core.convert.ElasticsearchTypeMapper;
+import org.springframework.data.elasticsearch.core.mapping.ElasticsearchPersistentEntity;
+import org.springframework.data.elasticsearch.core.mapping.ElasticsearchPersistentProperty;
+import org.springframework.data.mapping.MappingException;
+import org.springframework.data.mapping.PersistentPropertyAccessor;
+import org.springframework.data.mapping.context.MappingContext;
+import org.springframework.data.mapping.model.ConvertingPropertyAccessor;
+import org.springframework.data.mapping.model.PersistentEntityParameterValueProvider;
+import org.springframework.data.mapping.model.PropertyValueProvider;
+import org.springframework.data.util.ClassTypeInformation;
+import org.springframework.data.util.Streamable;
+import org.springframework.data.util.TypeInformation;
+import org.springframework.format.datetime.DateFormatterRegistrar;
+import org.springframework.lang.Nullable;
+import org.springframework.util.ClassUtils;
+import org.springframework.util.ObjectUtils;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+/**
+ * Elasticsearch specific {@link EntityReader} & {@link EntityWriter} implementation based on domain type
+ * {@link ElasticsearchPersistentEntity metadata}.
+ *
+ * @author Christoph Strobl
+ * @since 3.2
+ */
+public class ElasticsearchEntityMapper
+		implements EntityConverter<ElasticsearchPersistentEntity<?>, ElasticsearchPersistentProperty, Object, Map>,
+		EntityWriter<Object, Map>, EntityReader<Object, Map>, InitializingBean, EntityMapper {
+
+	private final MappingContext<? extends ElasticsearchPersistentEntity<?>, ElasticsearchPersistentProperty> mappingContext;
+	private final GenericConversionService conversionService;
+
+	private CustomConversions conversions = new ElasticsearchCustomConversions(Collections.emptyList());
+	private EntityInstantiators instantiators = new EntityInstantiators();
+
+	private ElasticsearchTypeMapper typeMapper;
+
+	public ElasticsearchEntityMapper(
+			MappingContext<? extends ElasticsearchPersistentEntity<?>, ElasticsearchPersistentProperty> mappingContext,
+			@Nullable GenericConversionService conversionService) {
+
+		this.mappingContext = mappingContext;
+		this.conversionService = conversionService != null ? conversionService : new DefaultConversionService();
+		this.typeMapper = ElasticsearchTypeMapper.defaultTypeMapper(mappingContext);
+	}
+
+	// --> READ
+
+	@Override
+	public <T> T readObject(Map<String, Object> source, Class<T> targetType) {
+		return read(targetType, source);
+	}
+
+	@Override
+	@Nullable
+	public <R> R read(Class<R> type, Map source) {
+		return doRead(source, ClassTypeInformation.from((Class<R>) ClassUtils.getUserClass(type)));
+	}
+
+	@Nullable
+	protected <R> R doRead(Map source, TypeInformation<R> typeHint) {
+
+		if (source == null) {
+			return null;
+		}
+
+
+
+		if(typeHint.getType().isInterface()) {
```

Comment:
Why is this happening only for interfaces? Wouldn't that be also useful for subtypes?